core/bootstrap/src/main/java/org/eclipse/dataspaceconnector/system/health/HealthCheckServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class HealthCheckServiceImpl implements HealthCheckService {
+    private final List<LivenessProvider> livenessProviders;
+    private final List<ReadinessProvider> readinessProviders;
+    private final List<StartupStatusProvider> startupStatusProviders;
+
+    private final Map<LivenessProvider, HealthCheckResult> cachedLivenessResults;
+    private final Map<ReadinessProvider, HealthCheckResult> cachedReadinessResults;
+    private final Map<StartupStatusProvider, HealthCheckResult> cachedStartupStatus;
+
+    private final ScheduledExecutorService executor;
+
+    public HealthCheckServiceImpl(HealthCheckServiceConfiguration configuration) {
+        readinessProviders = new CopyOnWriteArrayList<>();
+        livenessProviders = new CopyOnWriteArrayList<>();
+        startupStatusProviders = new CopyOnWriteArrayList<>();
+
+        cachedLivenessResults = new ConcurrentHashMap<>();
+        cachedReadinessResults = new ConcurrentHashMap<>();
+        cachedStartupStatus = new ConcurrentHashMap<>();
+
+        executor = Executors.newScheduledThreadPool(configuration.getThreadPoolSize());
+        start(configuration);
+    }
+
+    @Override
+    public void addLivenessProvider(LivenessProvider provider) {
+        livenessProviders.add(provider);
+    }
+
+    @Override
+    public void addReadinessProvider(ReadinessProvider provider) {
+        readinessProviders.add(provider);
+    }
+
+    @Override
+    public void addStartupStatusProvider(StartupStatusProvider provider) {
+        startupStatusProviders.add(provider);
+    }
+
+    @Override
+    public HealthStatus isLive() {
+        return new HealthStatus(cachedLivenessResults.values());
+    }
+
+    @Override
+    public HealthStatus isReady() {
+        return new HealthStatus(cachedReadinessResults.values());
+    }
+
+    @Override
+    public HealthStatus getStartupStatus() {
+        return new HealthStatus(cachedStartupStatus.values());
+    }
+
+    public void stop() {
+        if (!executor.isShutdown()) {
+            executor.shutdownNow();
+        }
+    }
+
+
+    private void start(HealthCheckServiceConfiguration configuration) {
+        //todo: maybe providers should provide their desired timeout instead of a global config?
+        executor.scheduleAtFixedRate(this::queryReadiness, 0, configuration.getReadinessPeriod().toMillis(), TimeUnit.MILLISECONDS);
+        executor.scheduleAtFixedRate(this::queryLiveness, 0, configuration.getLivenessPeriod().toMillis(), TimeUnit.MILLISECONDS);
+        executor.scheduleAtFixedRate(this::queryStartupStatus, 0, configuration.getStartupStatusPeriod().toMillis(), TimeUnit.MILLISECONDS);
+    }
```

Comment:
@paullatzelsperger Aren't those checks triggered from outside e.g. via a request handling thread / resp. why are these 3 extra threads required? Wouldn't caching results for a certain amount of time be sufficient?

Reply:
So the reasoning for this was that the probe requests coming in from K8S or docker will only hit the cached results and ttherefore be very fast, at the cost of potentially returning "stale" status. 

The service then periodically performs it's health checks separately, therefore decoupling the two mechanisms and avoiding system clogging: imagine we have a lot of backend systems, and every `readiness` probe triggers a heart beat to every one of them - that would certainly cause CPU load spikes or let the probes run into a timeout.

The reason why it's 3 threads as opposed to one is, that we might want to schedule them at different intervals.

Reply:
Okay.

What happens if a exception is thrown by any of the providers? - The spawning thread then dies and never gets restarted, right? Means either stale information is served fromt that point on or - if never run - the check never gets taken into consideration at all. Also `Stream#parallelStream` spawns another thread from the default `ForkJoinPool` for each element in the stream - seems a little bit wasteful. If you still want to keep the parallel stream execution please make it configurable which `ForkJoinPool` to use.
Btw, when latency of fulfilling a certain check exceeds the configured fixed delay there might be overlapping runs of that specific check.

Reply:
> What happens if a exception is thrown by any of the providers? - The spawning thread then dies and never gets restarted, right? Means either stale information is served fromt that point on - if ever run - or the check never gets taken into consideration at all.

good point. added exception handling.

> Also Stream#parallelStream spawns another thread from the default ForkJoinPool for each element in the stream - seems a little bit wasteful. If you still want to keep the parallel stream execution please make it configurable which ForkJoinPool to use.

The `threadPoolSize` of the `ScheduledExecutorService` is configurable, and the `parallelStreams` run within it's context, so I think we're good there. 

> Btw, when latency of fulfilling a certain check exceeds the configured fixed delay there might be overlapping runs of that specific check.

Yup, if heartbeats take longer than the interval configured in code, that can potentially create an ever-growing backlog of tasks. The alternative would be to keep track of running checks and only re-submit the check, if it's not already running. That will add more complexity and will still not prevent configuration errors, it would make them loggable though. Let me see if I can come up with a usable solution for that.

Reply:
Just to amend the parallelism topic: I did some testing, and from a runtime perspective it does not make a difference whether `new ForkJoinPool(5)` or a `threadPoolSize(5)` is used. Also, ForkJoinPools attempt to re-use internal worker threads, so it's not wasteful. 
Since health providers potentially do stuff that may take some time, having them checked in multiple threads in parallel is advisable here.